### ISSUE TITLE
<li>s contain <p> by default

### ISSUE THREAD

**Issue by arcataroger** [datocms/structured-text#18]
It seems like list items created in the editor have a `<p>` child by default:
![image](https://user-images.githubusercontent.com/11964962/125823926-2ef439d7-964a-4c1d-9488-f223c4659728.png)

Becomes:

```html
<ul>
    <li><p>This is a test</p></li>
    <li><p>Is this a test?</p></li>
    <li><p>A test am I?</p></li>
</ul>
```

This can break CSS and also the [renderRule for IsListItem](https://github.com/datocms/structured-text/blob/main/packages/generic-html-renderer/src/index.ts#L99-L101), because it gets forced to inherit the renderRule for `isParagraph`.

In our case we want to apply a bootstrap column to `isParagraph` renderRules, but then all our list items get forced to have columns as well.

I believe the `<li>` is not structurally required to have a `<p>` as a child (it can just contain text).

**Comment by arcataroger** [comment]
Actually, thinking about it, this might not be an issue with the `structured-text` module but somewhere upstream while the DAST tree is being created by the DatoCMS editor. If this is in the wrong please, please feel free to close it and let me know where to better file this. Thanks and sorry!

**Comment by stefanoverna** [comment]
Yes, as [per specs](https://www.datocms.com/docs/structured-text/dast#listItem) list items can only contain `paragraph` or `list` nodes, not directly `span`s. If you only want to tweak the rendering of top-level paragraphs, you can set up a renderRule that checks for ancestors:

Demo: https://stackblitz.com/edit/react-ts-2lqnxw?file=index.tsx

```js
<StructuredText
  data={data}
  customRules={[
    renderRule(
      isParagraph,
      ({ adapter: { renderNode }, node, children, key, ancestors }) => {
        if (isRoot(ancestors[0])) {
          // apply bootstrap column
          return renderNode('p', { key, className: 'bootstrap' }, children);
        } else {
          // default paragraph rendering...
          // return renderNode('p', { key }, children);

          // or even completely remove the paragraph and directly render the inner children:
          return children;
        }
      }
    )
  ]}
/>
```

**Comment by arcataroger** [comment]
I'm not sure how the DAST specs evolve, but given that [regular HTML `<li>`s don't HAVE to have a `<p>` inside them](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/li), is this something that could be considered/modified in the next update, perhaps? Maybe it's not a bug per se, but it seems a bit idiosyncratic that DAST forces every `<li>` to have a `<p>` when HTML doesn't require it.

Semantically, a lot of the content that we're going to be importing uses `<li>`s with a direct child (no `<p>`), and having the unnecessary one added by Dato means the stylesheets have to be modified along with any selectors and VDOM/JSX templates.

In the meantime, having access to the `ancestors` trait in `renderRule` is a great tip, thanks! I'll update that in the documentation :)

**Comment by arcataroger** [comment]
Unfortunately, it doesn't look like `ancestors` traverses all the way up to the root from a paragraph. For a `renderRule(isParagraph[...])` of a top-level paragraph, the `ancestors` object looks like this:

```json
[
  {
    "type": "paragraph",
    "children": [
      {
        "type": "span",
        "value": "The Field is open 9am to 5pm, Thursday through Monday. We’re currently closed on Tuesdays and Wednesdays. "
      }
    ]
  }
]
```

For an embedded paragraph inside an `<li>`, it's similar:
```json
[
  {
    "type": "list",
    "style": "bulleted",
    "children": [
      {
        "type": "listItem",
        "children": [
          {
            "type": "paragraph",
            "children": [
              {
                "type": "span",
                "value": "144 hand sanitizer stations are installed throughout the building"
              }
            ]
          }
        ]
      },
      {
        "type": "listItem",
        "children": [
          {
            "type": "paragraph",
            "children": [
              {
                "type": "span",
                "value": "Frequent, rigorous cleaning procedures follow CDC guidelines"
              }
            ]
          }
        ]
      },
      {
        "type": "listItem",
        "children": [
          {
            "type": "paragraph",
            "children": [
              {
                "type": "span",
                "value": "Floor markers help guide social distancing and direct one-way paths"
              }
            ]
          }
        ]
      },
      {
        "type": "listItem",
        "children": [
          {
            "type": "paragraph",
            "children": [
              {
                "type": "span",
                "value": "Exhibitions that are highly interactive or located in small spaces are temporarily closed"
              }
            ]
          }
        ]
      }
    ]
  },
  {
    "type": "listItem",
    "children": [
      {
        "type": "paragraph",
        "children": [
          {
            "type": "span",
            "value": "144 hand sanitizer stations are installed throughout the building"
          }
        ]
      }
    ]
  },
  {
    "type": "paragraph",
    "children": [
      {
        "type": "span",
        "value": "144 hand sanitizer stations are installed throughout the building"
      }
    ]
  }
]
```

`ancestors` goes all the way up to the `list` type but never all the way up to the `root`.

**Comment by stefanoverna** [comment]
Maybe you're using an older version of the package?

—
Stefano Verna
Founder and CEO
DatoCMS

--- original message ---
On July 16, 2021 at 6:15 PM GMT+2 ***@***.*** wrote:



Unfortunately, it doesn't look like ancestors traverses all the way up to the root from a paragraph. For a renderRule(isParagraph[...]) of a top-level paragraph, the ancestors object looks like this:

[
{
"type": "paragraph",
"children": [
{
"type": "span",
"value": "The Field is open 9am to 5pm, Thursday through Monday. We’re currently closed on Tuesdays and Wednesdays. "
}
]
}
]


For an embedded paragraph inside an <li>, it's similar:

[
{
"type": "list",
"style": "bulleted",
"children": [
{
"type": "listItem",
"children": [
{
"type": "paragraph",
"children": [
{
"type": "span",
"value": "144 hand sanitizer stations are installed throughout the building"
}
]
}
]
},
{
"type": "listItem",
"children": [
{
"type": "paragraph",
"children": [
{
"type": "span",
"value": "Frequent, rigorous cleaning procedures follow CDC guidelines"
}
]
}
]
},
{
"type": "listItem",
"children": [
{
"type": "paragraph",
"children": [
{
"type": "span",
"value": "Floor markers help guide social distancing and direct one-way paths"
}
]
}
]
},
{
"type": "listItem",
"children": [
{
"type": "paragraph",
"children": [
{
"type": "span",
"value": "Exhibitions that are highly interactive or located in small spaces are temporarily closed"
}
]
}
]
}
]
},
{
"type": "listItem",
"children": [
{
"type": "paragraph",
"children": [
{
"type": "span",
"value": "144 hand sanitizer stations are installed throughout the building"
}
]
}
]
},
{
"type": "paragraph",
"children": [
{
"type": "span",
"value": "144 hand sanitizer stations are installed throughout the building"
}
]
}
]


ancestors goes all the way up to the list type but never all the way up to the root.

—

You are receiving this because you commented.

Reply to this email directly, view it on GitHub, or unsubscribe.
--- end of original message ---

**Comment by arcataroger** [comment]
Hmm, I don't think so? `"react-datocms": "^1.6.3"`

**Comment by stefanoverna** [comment]
Sorry for the delay, I took a bit of vacation..  it's datocms-structured-text-utils that needs to be upgraded!

**Comment by arcataroger** [comment]
No worries, glad you had a vacation :)

We don't use `datocms-structured-text-utils` directly as a dependency. But it's included in `react-datocms` @ ^1.1.0, while the latest is 1.2.0. Maybe it needs to be upgraded there?

This is not an urgent issue for us; we've switched to using Modular Content Fields instead of Structured Text. But just thought I'd mention it. Thanks!

**Comment by stefanoverna** [comment]
Yes, that's correct, you should upgrade the deep dependency.

—
Stefano Verna
Founder and CEO
DatoCMS

--- original message ---
On July 20, 2021 at 4:28 PM GMT+2 ***@***.*** wrote:



No worries, glad you had a vacation :)

We don't use datocms-structured-text-utils directly as a dependency. But it's included in react-datocms @ ^1.1.0, while the latest is 1.2.0. Maybe it needs to be upgraded there?

This is not an urgent issue for us; we've switched to using Modular Content Fields instead of Structured Text. But just thought I'd mention it. Thanks!

—

You are receiving this because you commented.

Reply to this email directly, view it on GitHub, or unsubscribe.
--- end of original message ---
